### PR TITLE
Phase2-hgx137 Modify the code to go toward the DIGI step for HFNose

### PR DESCRIPTION
**Pull request by bsunanda**
Utilizer the codes for HGCal EE

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: SimCalorimetry/HGCalSimProducers/python/hgcalDigitizer_cfi.py [hunk 1]
```diff
@@ -191,6 +191,72 @@
...
+                 -21.85, 49.39,  22.21,  0.8,
+                 -0.28,   27.14,  43.95,
+                 3.89048 )
+            )
+        )
+    )
+
 # this bypasses the noise simulation
 from Configuration.ProcessModifiers.premix_stage1_cff import premix_stage1
 for _m in [hgceeDigitizer, hgchefrontDigitizer, hgchebackDigitizer]:
```
Comment (makortel) [inline SimCalorimetry/HGCalSimProducers/python/hgcalDigitizer_cfi.py]: By looking the code I'd expect premixing work out of the box, except this loop needs `hfnoseDigitizer`.

Comment (bsunanda) [reply SimCalorimetry/HGCalSimProducers/python/hgcalDigitizer_cfi.py]: Thanks - I am adding that now

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
@bsunanda there are a few additional things to do here:
1. make new Eras `Phase2C6` and `Phase2C6_timing` (in `Configuration/Eras/python`) that add `phase2_hfnose` to `Phase2C4` and `Phase2C4_timing` (respectively)
2. update [upgradeWorkflowComponents.py](https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py) to use `Phase2C6`
3. Update [Configuration/Geometry/README.md](https://github.com/cms-sw/cmssw/blob/master/Configuration/Geometry/README.md) with the new subdetector versions

**Comment by kpedro88** [comment]
@bsunanda please add the new Eras in `Configuration/StandardSequences/python/Eras.py`

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@kpedro88 Please test

File: SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc [hunk 2]
```diff
@@ -582,6 +602,9 @@ uint32_t HGCDigitizer::getType() const {
     case DetId::HGCalHSc:
       idx = 2;
       break;
+    case DetId::Forward:
+      idx = 0;
```
Comment (kpedro88) [inline SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc]: does it make sense to assign index 0 to both EE and HFNose?

File: SimGeneral/PreMixingModule/python/mixOne_premix_on_sim_cfi.py [hunk 3]
```diff
@@ -337,6 +338,18 @@
     )
 )
 
+phase2_hfnose.toModify(mixData,
```
Comment (kpedro88) [inline SimGeneral/PreMixingModule/python/mixOne_premix_on_sim_cfi.py]: @makortel FYI, we should check if the premixing setup works out of the box for this new detector (an HGCal-like section in front of HF)

Comment (makortel) [reply SimGeneral/PreMixingModule/python/mixOne_premix_on_sim_cfi.py]: @kpedro88 From visual inspection I'd expect so (but of course it should be tested). Does this require the v9 geometry of HGCal (for which the premixing still needs to be fixed #23624)?

File: SimG4CMS/Calo/interface/HFNoseSD.h [hunk 4]
```diff
@@ -44,7 +44,7 @@ class HFNoseSD : public CaloSD, public Observer<const BeginOfJob *> {
   const HGCalDDDConstants*               hgcons_;
   std::unique_ptr<HFNoseNumberingScheme> numberingScheme_;
   std::unique_ptr<HGCMouseBite>          mouseBite_;
-  const std::string                      nameX_ = "HFNoseSensitive";
+  const std::string                      nameX_ = "HGCalHFNoseSensitive";
```
Comment (civanch) [inline SimG4CMS/Calo/interface/HFNoseSD.h]: @bsunanda , I would say it is unnecessary overhead to add nameX_, there is no need to have two names - one from GetName() another from local variable. I would suggest to drop this class member.

File: SimGeneral/PreMixingModule/python/mixOne_premix_on_sim_cfi.py (hunk 3 above)
Comment (kpedro88) [reply SimGeneral/PreMixingModule/python/mixOne_premix_on_sim_cfi.py]: So far the HFNose is only in a Phase2 geometry with HGCal v9.

Comment (makortel) [reply SimGeneral/PreMixingModule/python/mixOne_premix_on_sim_cfi.py]: Ok, se we need to fix #23624 first.

**Comment by ianna** [comment]
+1

File: SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc (hunk 2 above)
Comment (bsunanda) [reply SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc]: Changed it to 4

File: SimG4CMS/Calo/interface/HFNoseSD.h (hunk 4 above)
Comment (bsunanda) [reply SimG4CMS/Calo/interface/HFNoseSD.h]: It is the name of the Geometry object. We cannot get that from GetName().

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: SimG4CMS/Calo/src/HFNoseSD.cc [hunk 5]
```diff
@@ -47,6 +47,9 @@ HFNoseSD::HFNoseSD(const std::string& name, const DDCompactView & cpv,
   waferRot_        = m_HFN.getParameter<bool>("RotatedWafer");
   angles_          = m_HFN.getUntrackedParameter<std::vector<double>>("WaferAngles");
 
+  nameX_ = ((name.find("HFNoseHits")!=std::string::npos) ? 
+	    "HGCalHFNoseSensitive" : "HFNoseSensitive");
+
```
Comment (civanch) [inline SimG4CMS/Calo/src/HFNoseSD.cc]: @bsunanda , I may be do not understand something but why sensitive detector name has "HFNoseHits" name? name is copied to SD name, which is accessible via GetName() top method. If we recognise DetID via name, this name can be used. If not, I would instead use enumeration and not strings nameX_.

Comment (bsunanda) [reply SimG4CMS/Calo/src/HFNoseSD.cc]: name or GetName() returns the name of the hit collection and not the name of SD.

Comment (bsunanda) [reply SimG4CMS/Calo/src/HFNoseSD.cc]: This has been the convention in CMS from the beginning

**Comment by bsunanda** [comment]
@kpedro88 I cannot get the test 26234.0. Do I need to merge some additional PR to run this workflow?

**Comment by bsunanda** [comment]
@kpedro88 The fix is in #24264. Once that is integrated the D31 will run correctly

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by ianna** [comment]
+1

File: SimG4CMS/Calo/src/HFNoseSD.cc (hunk 5 above)
Comment (civanch) [reply SimG4CMS/Calo/src/HFNoseSD.cc]: @bsunanda , from what I observed in the code by fact there was no difference between these two.

**Comment by civanch** [comment]
+1

**Comment by kpedro88** [comment]
please test with #24264

**Comment by bsunanda** [comment]
@kpedro88 That test will fail with this PR - it will pass only after PR #24264 is merged. I cannot put the same fix in 2 PR's

**Comment by kpedro88** [comment]
@bsunanda I am testing those two PRs together

**Comment by kpedro88** [comment]
please test with #24264

**Comment by kpedro88** [comment]
+upgrade

**Comment by bsunanda** [comment]
@kpedro88 @fabiocos @franzoni Please sign and merge this PR. I have other code which I would like to add after this is merged

**Comment by kpedro88** [comment]
@fabozzi @prebello please sign

**Comment by bsunanda** [comment]
@fabozzi @prebello please sign
@kpedro88 @fabiocos Please merge

**Comment by kpedro88** [comment]
@fabozzi @prebello @GurpreetSinghChahal @zhenhu @pgunnell ping

**Comment by fabozzi** [comment]
+1

**Comment by kpedro88** [comment]
+1

**Comment by kpedro88** [comment]
merge

**Pull request merged**